daemon/src/engine/framework/Rcon.cpp

```diff
@@ -0,0 +1,252 @@
+    2
+);
+
+Cvar::Cvar<std::string> cvar_client_destination(
+    "rcon.client.destination",
+    "Destination address for rcon commands, if empty the current server.",
+    Cvar::NONE,
+    ""
+);
+
+Message::Message( const netadr_t& remote, std::string command,
+				  Secure secure, std::string password, std::string challenge )
+	: secure_(secure),
+	  challenge_(std::move(challenge)),
+	  command_(std::move(command)),
+	  password_(std::move(password)),
+	  remote_(remote)
+{}
+
+Message::Message( std::string error_message )
+	: error_(std::move(error_message))
+{}
+
+const std::string& Message::command() const
+{
+	return command_;
+}
+
+void Message::send() const
+{
+	if ( secure_ == Secure::Unencrypted )
+	{
+		Net::OutOfBandPrint(netsrc_t::NS_CLIENT, remote_, "rcon %s %s", password_, command_);
+		return;
+	}
+
+	std::string method = "PLAIN";
+	Crypto::Data key = Crypto::Hash::Sha256(Crypto::String(password_));
+	std::string plaintext = command_;
+
+	if ( secure_ == Secure::EncryptedChallenge )
+	{
+		method = "CHALLENGE";
+		plaintext = challenge_ + ' ' + plaintext;
+	}
+
+	Crypto::Data cypher;
+	if ( Crypto::Aes256Encrypt(Crypto::String(plaintext), key, cypher) )
+	{
+		Net::OutOfBandPrint(netsrc_t::NS_CLIENT, remote_, "srcon %s %s",
+			method,
+			Crypto::String(Crypto::Encoding::Base64Encode(cypher))
+		);
+	}
+}
+
+bool Message::valid(std::string *invalid_reason) const
+{
+	auto invalid = [invalid_reason](const char* reason)
+	{
+		if ( invalid_reason )
+			*invalid_reason = reason;
+		return false;
+	};
+
+	if ( !error_.empty() )
+	{
+		return invalid(error_.c_str());
+	}
+
+	if ( secure_ < Secure::Unencrypted || secure_ > Secure::EncryptedChallenge )
+	{
+		return invalid("Unknown secure protocol");
+	}
+
+	if ( password_.empty() )
+	{
+		return invalid("Missing password");
+	}
+
+	if ( command_.empty() )
+	{
+		return invalid("Missing command");
+	}
+
+	if ( secure_ == Secure::EncryptedChallenge && challenge_.empty() )
+	{
+		return invalid("Missing challenge");
+	}
+
+	return true;
+}
+
+
+bool Message::acceptable(std::string *invalid_reason) const
+{
+	auto invalid = [invalid_reason](const char* reason)
+	{
+		if ( invalid_reason )
+			*invalid_reason = reason;
+		return false;
+	};
+
+	if ( !valid(invalid_reason) )
+	{
+		return false;
+	}
+
+	if ( secure_ < Secure(cvar_server_secure.Get()) )
+	{
+		return invalid("Weak security");
+	}
+
+	if ( cvar_server_password.Get().empty() )
+	{
+		return invalid("No rcon.server.password set on the server.");
+	}
+
+	if ( password_ != cvar_server_password.Get() )
+	{
+		return invalid("Bad password");
+	}
+
+	if ( secure_ == Secure::EncryptedChallenge )
+	{
+		if ( !ChallengeManager::Get().MatchString(remote_, challenge_) )
+		{
+			return invalid("Mismatched challenge");
+		}
+	}
+
+	return true;
+}
+
+Message Message::decode(const netadr_t& remote, const Cmd::Args& args)
+{
+	if ( args.size() < 3 || (args[0] != "rcon" && args[0] != "srcon") )
+	{
+		return Message("Invalid command");
```

Comment:
Sounds kinda ugly to return the same class for both error and success. This might be clearer if you use something like Util::Optional.